sys/shell/shell.c

```diff
@@ -66,160 +60,195 @@
+static shell_command_handler_t find_handler(
+        const shell_command_t *command_list, char *command)
 {
-    printf("%-20s %s\n", "Command", "Description");
-    puts("---------------------------------------");
+    shell_command_handler_t handler = NULL;
+    if (command_list != NULL) {
+        handler = search_commands(command_list, command);
+    }
+
+    if (handler == NULL && _builtin_cmds != NULL) {
+        handler = search_commands(_builtin_cmds, command);
+    }
 
-    const shell_command_t *command_lists[] = {
-        command_list,
-        MORE_COMMANDS
-    };
+    return handler;
+}
 
-    /* iterating over command_lists */
-    for (unsigned int i = 0; i < ARRAY_SIZE(command_lists); i++) {
+static void print_commands(const shell_command_t *entry)
+{
+    for (; entry->name != NULL; entry++) {
+        printf("%-20s %s\n", entry->name, entry->desc);
+    }
+}
 
-        const shell_command_t *entry;
+static void print_help(const shell_command_t *command_list)
+{
+    puts("Command              Description"
+         "\n---------------------------------------");
+    if (command_list != NULL) {
+        print_commands(command_list);
+    }
 
-        if ((entry = command_lists[i])) {
-            /* iterating over commands in command_lists entry */
-            while (entry->name != NULL) {
-                printf("%-20s %s\n", entry->name, entry->desc);
-                entry++;
-            }
-        }
+    if (_builtin_cmds != NULL) {
+        print_commands(_builtin_cmds);
     }
 }
 
+/**
+ * Break input line into words, create argv and call the command handler.
+ *
+ * Words are broken up at spaces. A backslash escaped the character that comes
+ * after (meaning if it is taken literally and if it is a space it does not break
+ * the word). Spaces can also be protected by quoting with double or single
+ * quotes.
+ *
+ State diagram for the tokenizer:
+```
+           ┌───[\]────┐   ┌─────["]────┐   ┌───[']─────┐  ┌───[\]────┐
+           ↓          │   ↓            │   │           ↓  │          ↓
+  ┏━━━━━━━━━━┓      ┏━┷━━━━━┓        ┏━┷━━━┷━┓       ┏━━━━┷━━┓     ┏━━━━━━━━━━┓
+  ┃DQUOTE ESC┃      ┃DQUOTE ┠───["]─>┃SPACE  ┃<─[']──┨SQUOTE ┃     ┃SQUOTE ESC┃
+  ┗━━━━━━━━┯━┛      ┗━━━━━━┯┛        ┗┯━━━━┯━┛       ┗━┯━━━━━┛     ┗━━━┯━━━━━━┛
+           │         ↑     │          │    │           │     ↑(store)  │
+           │  (store)│     │   ┌─[\]──┘    └──[*]────┐ │     │         │
+           └──[*]──▶┴◀[*]┘   │                     │ └[*]▶┴◀──[*]──┘
+                               ↓     ┏━━━━━━━┓       ↓
+                               ├◀[\]┨NOQUOTE┃◀─────┼◀─┐
+                               │     ┗━━━━━┯━┛(store)↑   │
+                               │           │         │   │
+                               │           └─[*]─────┘   │
+                               │     ┏━━━━━━━━━━━┓       │
+                               └───▶┃NOQUOTE ESC┠──[*]──┘
+                                     ┗━━━━━━━━━━━┛
```

Comment:
```suggestion
           ┌───[\]────┐   ┌─────["]────┐   ┌───[']─────┐  ┌───[\]────┐
           ↓          │   ↓            │   │           ↓  │          ↓
  ┏━━━━━━━━━━┓      ┏━┷━━━━━┓        ┏━┷━━━┷━┓       ┏━━━━┷━━┓     ┏━━━━━━━━━━┓
  ┃DQUOTE ESC┃      ┃DQUOTE ┠───["]─>┃SPACE  ┃<─[']──┨SQUOTE ┃     ┃SQUOTE ESC┃
  ┗━━━━━━━━┯━┛      ┗━━━━━━┯┛        ┗┯━━━━┯━┛       ┗━┯━━━━━┛     ┗━━━┯━━━━━━┛
           │         ↑     │          │    │           │     ↑(store)  │
           │  (store)│     │   ┌─[\]──┘    └──[*]────┐ │     │         │
           └──[*]───▶┴◀─[*]┘   │                     │ └[*]▶┴◀──[*]────┘
                               ↓     ┏━━━━━━━┓       ↓
                               ├◀[\]─┨NOQUOTE┃◀──────┼─◀─┐
                               │     ┗━━━━━┯━┛(store)↑   │
                               │           │         │   │
                               │           └─[*]─────┘   │
                               │     ┏━━━━━━━━━━━┓       │
                               └────▶┃NOQUOTE ESC┠──[*]──┘
                                     ┗━━━━━━━━━━━┛
```

Reply:
I checked this and I realized that I can't format this graph in a way such that it looks good on my terminal and my IDE at the same time. Right now it looks good for my IDE, but on the terminal it's broken. What would you suggest in this case? Add an image instead or optimize it for one of these programs?

Reply:
Whats your IDE? By in the terminal do you mean in `vim` or something like that? For me it looks the same in `vscode`, `vim` , or `cat` the file

Reply:
I use CLion, which is IntelliJ based. Even though a monospace font is setup by default, the black triangle character seems to be larger than the other characters, therefore the issue in these lines. I will push a fixup commit which includes a comment about this in the code.